### PR TITLE
Parameter management utilities

### LINKED ISSUE #96: Comprehensive way to manage ReSolve configuration parameters
ReSolve includes several different linear solvers and supports three hardware backends. There is a large number of parameters in ReSolve to select specific solution strategy and to tune its performance. Many of the parameters are specific to a specific solver type or hardware backend or both. We need a hierarchical method to allow user to configure ReSolve easily without sifting through dozens of (often cryptically described) parameters. Most of these parameters need to be set at runtime.

This issue is to start discussion how to handle this most effectively.

The issue originally raised by @kswirydo in https://github.com/ORNL/ReSolve/pull/86#discussion_r1412715592_

### PR DESCRIPTION
**Pull request by pelesh**
Managing different solver parameters in Re::Solve becomes a nontrivial task and there is a need for consistent parameter management across all Re::Solve modules. In addition to setting up file I/O for initializing solver parameters, the scope here includes creating a common interface for setting and retrieving parameters for all solver classes and ability to change solver configuration parameters at runtime. The requirements are:
- [x] Create a generic parameter management interface at the base class, which each derived solver class can populate with its own specific parameters.
- [x] Parameters can be of different types (integer, floating point, string, etc.).
- [x] This is core functionality, no additional dependency should be required to implement it.
- [x] Changing parameters should leave solver in fully functional state with new parameter values.

This PR is a proposal how to handle this. Parameter management is implemented only for classes `LinSolverIterativeFGMRES` and `LinSolverIterativeRandGMRES`. After this is agreed upon and merged, parameter management will be added to other solvers.

### CODE REVIEW

File: resolve/LinSolver.hpp [hunk 1]
```diff
@@ -34,94 +48,44 @@ namespace ReSolve
       virtual ~LinSolver();
 
       real_type evaluateResidual();
+
+      virtual int setCliParam(const std::string /* id */, const std::string /* value */)
+      {
+        return 1;
+      }
```
Comment (cameronrutherford) [inline resolve/LinSolver.hpp:55]: I have a personal preference for these functions being one line, but that has been discussed in #151

File: tests/functionality/testSysRandGMRES.cpp [hunk 2]
```diff
@@ -129,17 +129,23 @@ int test(int argc, char *argv[])
   real_type norm_b = 0.0;
 
   // Set solver options
-  real_type tol = 1e-12;
+  //real_type tol = 1e-12;
```
Comment (cameronrutherford) [inline tests/functionality/testSysRandGMRES.cpp:132]:
```suggestion
```

File: tests/functionality/testSysRandGMRES.cpp [hunk 3]
```diff
@@ -129,17 +129,23 @@ int test(int argc, char *argv[])
...
   solver.getIterativeSolver().setMaxit(2500);
-  solver.getIterativeSolver().setTol(tol);
+  // solver.getIterativeSolver().setTol(tol);
```
Comment (cameronrutherford) [inline tests/functionality/testSysRandGMRES.cpp:134]:
```suggestion
```

File: tests/functionality/testSysRandGMRES.cpp [hunk 4]
```diff
@@ -129,17 +129,23 @@ int test(int argc, char *argv[])
...
+  solver.getIterativeSolver().getCliParam("tol", tol);
 
   matrix_handler.setValuesChanged(true, memspace);
 
   // Set system matrix and initialize iterative solver
   status = solver.setMatrix(A);
   error_sum += status;
 
-  solver.getIterativeSolver().setRestart(200);
+  // solver.getIterativeSolver().setRestart(200);
```
Comment (cameronrutherford) [inline tests/functionality/testSysRandGMRES.cpp:145]:
```suggestion
```

File: resolve/LinSolverIterativeRandFGMRES.cpp [hunk 5]
```diff
@@ -502,6 +505,109 @@ namespace ReSolve
...
+  int LinSolverIterativeRandFGMRES::getCliParam(const std::string id, std::string& /* value */)
+  {
+    switch (params_list_[id])
+    {
+      default:
+        out::error() << "Unknown string parameter " << id << "\n";
+        return 1;
+    }
+    return 0;
+  }
```
Comment (cameronrutherford) [inline resolve/LinSolverIterativeRandFGMRES.cpp:541]: I am a little confused about the purpose of this function. Is that to say no parameters have a string as their value?

File: resolve/LinSolverIterativeFGMRES.cpp [hunk 6]
```diff
@@ -143,18 +146,31 @@ namespace ReSolve
...
+      //   exit_cond =  ((std::abs(rnorm - ZERO) <= EPSILON));
+      // } else {
+      //   if (conv_cond_ == 1) {
+      //     exit_cond =  ((std::abs(rnorm - ZERO) <= EPSILON) || (rnorm < tol_));
+      //   } else {
+      //     if (conv_cond_ == 2) {
+      //       exit_cond =  ((std::abs(rnorm - ZERO) <= EPSILON) || (rnorm < (tol_*bnorm)));
+      //     }
+      //   }
+      // }
```
Comment (cameronrutherford) [inline resolve/LinSolverIterativeFGMRES.cpp:173]: Should probably remove this if it's unnecessary as the switch statement is cleaner.

**Review by cameronrutherford — COMMENTED** [review]
I don't really have substantial feedback...

File: tests/functionality/testSysRandGMRES.cpp (hunk 2 above)
Comment (pelesh) [reply tests/functionality/testSysRandGMRES.cpp:132]: Good catch!

File: resolve/LinSolverIterativeRandFGMRES.cpp (hunk 5 above)
Comment (pelesh) [reply resolve/LinSolverIterativeRandFGMRES.cpp:541]: You are right. The base class should implement parameter getters returning error messages. Derived classes should implement parameter getters only for parameters that are defined for that specific solver.

File: resolve/LinSolverIterativeFGMRES.cpp (hunk 6 above)
Comment (pelesh) [reply resolve/LinSolverIterativeFGMRES.cpp:173]: Thanks!

**Comment by stonecoldhughes** [comment]
https://github.com/ORNL/ReSolve/issues/96#issuecomment-2362551756

**Comment by stonecoldhughes** [comment]
https://en.cppreference.com/w/cpp/language/raii is an idea that applies to configuration - ideally, an object is completely configured and ready to use as soon as it is constructed. Are these configurable solver objects configured with the parsed configuration options in their constructors? Or is the functionality here used to set that up after the fact?

**Comment by pelesh** [comment]
> https://en.cppreference.com/w/cpp/language/raii is an idea that applies to configuration - ideally, an object is completely configured and ready to use as soon as it is constructed. Are these configurable solver objects configured with the parsed configuration options in their constructors? Or is the functionality here used to set that up after the fact?

Since Re::Solve is used for solving series of similar linear systems it is conceivable that one might want to change its parameters at runtime. In that case, Re::Solve would need to ensure that the choice of solver configuration parameters is self-consistent and that correct workspace is allocated.

File: resolve/LinSolver.hpp (hunk 1 above)
Comment (pelesh) [reply resolve/LinSolver.hpp:55]: To be tracked in #151.

File: resolve/LinSolverIterativeFGMRES.cpp (hunk 6 above)
Comment (pelesh) [reply resolve/LinSolverIterativeFGMRES.cpp]: Removed in [5356946](https://github.com/ORNL/ReSolve/pull/163/commits/535694632e9ebc197475d5649ec249c310b134c1).

**Review by stonecoldhughes — APPROVED** [review]
These changes are a big step in the right direction. According to https://en.wikipedia.org/wiki/Single-responsibility_principle, it is usually best to map each separate definable functionality to a separately testable object. The LinSolver base class appears to handle 3 things: parameter management, norm/residual calculation, and matrix equation solution. Is it possible for LinSolver to offload the first two into other classes to a greater degree? Would that increase the composability of the design in a desirable way?

**Comment by pelesh** [comment]
I suggest we keep this as draft for now. We need to merge #198 and probably do some other fixes as we discover them through this preview PR. Thanks all for useful feedback so far.

File: tests/functionality/testSysRandGMRES.cpp (hunk 2 above)
Comment (pelesh) [reply tests/functionality/testSysRandGMRES.cpp]: Fixed in [8e1bd44](https://github.com/ORNL/ReSolve/pull/163/commits/8e1bd44561ae39bf6888da14d5c783681fe01bd2).

File: tests/functionality/testSysRandGMRES.cpp (hunk 4 above)
Comment (pelesh) [reply tests/functionality/testSysRandGMRES.cpp]: Fixed in [8e1bd44](https://github.com/ORNL/ReSolve/pull/163/commits/8e1bd44561ae39bf6888da14d5c783681fe01bd2).

File: resolve/LinSolver.hpp [hunk 7]
```diff
@@ -34,94 +48,44 @@ namespace ReSolve
...
+        
+      virtual int getCliParam(const std::string /* id */, std::string& /* value */)
+      {
+        return 1;
+      }
+        
+      virtual int getCliParam(const std::string /* id */, index_type& /* value */)
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/LinSolver.hpp:62]: They can and they probably should. In this PR, we implement parameter management only for a couple of solver classes, just to see if this design is feasible. Once all solver classes implement these methods, we'll turn base class methods pure virtual.

File: tests/functionality/testSysRandGMRES.cpp (hunk 3 above)
Comment (pelesh) [reply tests/functionality/testSysRandGMRES.cpp]: Fixed in [8e1bd44](https://github.com/ORNL/ReSolve/pull/163/commits/8e1bd44561ae39bf6888da14d5c783681fe01bd2).

**Comment by pelesh** [comment]
> https://en.cppreference.com/w/cpp/language/raii is an idea that applies to configuration - ideally, an object is completely configured and ready to use as soon as it is constructed. Are these configurable solver objects configured with the parsed configuration options in their constructors? Or is the functionality here used to set that up after the fact?

This is a very good point. In sparse numerical linear algebra, one needs to compute the size of the workspace _after_ all the objects are instantiated. Similarly, changing some of the configuration parameters requires recomputing and reallocating the workspace. Therefore, solver object typically is _not_ ready to use as soon as it is constructed.

File: tests/functionality/testSysRandGMRES.cpp [hunk 8]
```diff
@@ -177,7 +198,7 @@ int test(int argc, char *argv[])
     std::cout << "Result is not a finite number!\n";
     error_sum++;
   }
-  if (final_norm/norm_b > (10.0 * tol)) {
+  if (final_norm/norm_b > (10.0 * tol_out)) {
```
Comment (shakedregev) [inline tests/functionality/testSysRandGMRES.cpp:201]: Maybe this `10.0` should be an optional parameter? I can imagine a scenario where I specify a small tolerance, but don't necessarily want to throw an error if I wasn't able to achieve it (or 10X it)

**Review by shakedregev — APPROVED** [review]
I pulled and ran the new test, it works. The code is understandable, but we should be able to find a way to make these MRs more manageable, as much of the code was repeated.

File: tests/functionality/testSysRandGMRES.cpp (hunk 8 above)
Comment (pelesh) [reply tests/functionality/testSysRandGMRES.cpp:201]: See #203, for example. We do need to make these tests more modular and more configurable.

**Review by maksud — APPROVED** [review]
Looks good to me based on the discussions.

**Pull request merged**